### ISSUE TITLE
Browser not refreshing 

### ISSUE THREAD

**Issue by gyps** [ch11ng/exwm#814]
Dear all,

since a few weeks I have an issue with the window of my browser (i.e. chromium) not refreshing for about 30-60secs from time to time. This happens usually when switching back to the browser window. Another program I'm using had the same issue but in this case no refresh at all took place (not even after 1 hour). So the only solution is to kill the program. 

Obviously there are a lot of causes that can be the reason for this behavior but I have no clue where to start searching for a reason.

Does anybody observe similar behavior? Or has a hint where to start searching for a cause?

Thanks,
Alex

**Comment by gyps** [comment]
One more thing. If I switch from another workspace back to the workspace the browserwindow is in the window has focus but the application doesn't. I always have to click into the window to really focus into the program. Also this was not always liks this but changed within the last weeks. I don't know if this is related but it is a strange coincidence at least.

**Comment by tazjin** [comment]
I had this exact issue mysteriously start a few months ago, and stop again a few weeks later. I think there's a range of Chromium versions where this happens and eventually it stops.

**Comment by BloggerBust** [comment]
I have the same issue with the following applications: Chrome, Skype, Teams and most recently Firefox v 78.8.0esr. It never use to be a problem for me either. I found that I can always. eventually, get the UI to redraw by doing combinations of the following:

- Split window with C-x 3 and delete other window with C-x 1 a few times
- call redraw-display

I also found that signing into google accounts in chrome seemed to help.

Here are some observations:
- When running Chrome, Firefox, teams all at the same time, the problem might only impact a subset of these applications
- The problem use to be worse for chrome and teams, but recent updates has improved the situation. I am now running Chrome Version 89.0.4389.72 (Official Build) (64-bit)
- A recent update of Firefox has added it to the list of impacted applications
- The problem is not related to memory , CPU or disk I/O (I have seen the issue with low usage of memory, CPU and I/O)
- The problem happens every day at seemingly random times. but is sometimes much worse than other times.

**Comment by BloggerBust** [comment]
This is probably not an EXWM issue. I have found that whenever chrome UI becomes unresponsive, the following error appears in the *google-chrome-stable* buffer.

`[20914:20914:0312/053322.508703:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.`

@gyps Are you seeing the same error when chrome fails to redraw/refresh?

Today I will monitor teams and to see if the same error occurs.

**Comment by gyps** [comment]
Unfortunately the issue persists. Firefox as well as Chromium once in a while (usually after a few minutes) do not refresh any more when they receive focus. Some versions don't have this problem as often as others but currently it is again really annoying.  
I started the application from a terminal and when this happens the app is absolutely quite. No terminal output whatsoever. I'm absolutely clueless.

**Comment by BloggerBust** [comment]
@gyps Do you have a custom kernel? If so, can you check if `CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE=y`.

`grep -i CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE  /usr/src/linux/.config`

If your kernel is configured to require user space tools to govern CPU frequency scaling by default, but you have not installed or configured any user space tools to do so, then your systems performance will suffer. For some reason, my kernel was configured with `CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE=y`, so I changed it to `CONFIG_CPU_FREQ_DEFAULT_GOV_SCHEDUTIL=y` and I have noticed a significant improvement in GUI responsiveness and memory usage.

**Comment by gyps** [comment]
No, I'm not using a custom kernel. My current kernel is current arch-linux kernel 
5.12.14-arch1-1 #1 SMP PREEMPT Thu, 01 Jul 2021 07:26:06 +0000 x86_64 GNU/Linux
I've no idea what compile flags are set in this arch kernel. On the other hand - when I first encountered this issue I was still using Debian. Therefore I don't expect this to be the cause of the problem.

**Comment by Parinz** [comment]
Is there any progress on the issue? I think it might be GPU based. I have integrated graphics. What are you running?

For extra context, I am also using firefox.

**Comment by gyps** [comment]
Im also using integrated GPU on a DELL XPS-13 Laptop. Problem persists with firefox and chromium.

**Comment by Parinz** [comment]
I think this has a high chance of being a d-bus issue. Firefox seems to crash saying things about "Channel Error". When I searched that up it lead me to [this.](https://www.reddit.com/r/archlinux/comments/cx4h5a/firefox_keeps_crashing/) The person said that the problem was due to a misconfiguration in the gpu config of his coreboot.

So it definitely has something to do with the GPU and the d-bus. This is strange. I haven't seen this before. I'm currently running Arch Linux, I plan to switch to Gentoo. I'll update you whether if the same problem persists in Gentoo.

**Comment by BloggerBust** [comment]
Check to see if your [graphics card has gone legacy](https://www.nvidia.com/en-us/drivers/unix/legacy-gpu/). If it has, then only a fixed range of driver versions may be compatible with your card. If you have an nvidia graphics card, then you can use [this form](https://www.nvidia.com/Download/Find.aspx) to determine what versions of the driver your card is compatible with. You will then need to follow the instructions for your distributions package manager to mask your graphics card drivers to the latest compatible version. I suspect that at least some of you will find that your graphics card driver has gone legacy and that following these steps will resolve the refresh issue.

Please post back to tell us whether or not this fixes your issue.

**Comment by BloggerBust** [comment]
@Gyps, If you are unable to resolve this issue after reviewing [the arch-linux intel wiki](https://wiki.archlinux.org/title/Intel_graphics) and [trouble shooting section](https://wiki.archlinux.org/title/Intel_graphics#Troubleshooting), I recommend asking this question on [the arch-linux community forum](https://bbs.archlinux.org/) as I do not believe it is in anyway related to exwm. I have been able to completely resolve the issue on my laptop. The root cause in my case was that my graphics card went legacy and I ended up installing a version of the driver that did not support my graphics card.
